README.md
# flat-tree

[![crates.io version][1]][2] [![build status][3]][4]
[![downloads][5]][6] [![docs.rs docs][7]][8]

Map a binary tree to a list. Adapted from
[mafintosh/flat-tree](https://github.com/mafintosh/flat-tree).

- [Documentation][8]
- [Crates.io][2]

## License
[MIT](./LICENSE-MIT) OR [Apache-2.0](./LICENSE-APACHE)

[1]: https://img.shields.io/crates/v/flat-tree.svg?style=flat-square
[2]: https://crates.io/crates/flat-tree
[3]: https://github.com/datrs/flat-tree/actions/workflows/ci.yml/badge.svg
[4]: https://github.com/datrs/flat-tree/actions
[5]: https://img.shields.io/crates/d/flat-tree.svg?style=flat-square
[6]: https://crates.io/crates/flat-tree
[7]: https://docs.rs/flat-tree/badge.svg
[8]: https://docs.rs/flat-tree


CHANGELOG.md
## 2023-04-12, Version 6.0.0
### Commits

- [[`6a7bbb5272`](https://github.com/datrs/flat-tree/commit/6a7bbb5272b51754ce47ee12430c5221ad22aa7b)] Release 6.0.0 (Timo Tiuraniemi)
- [[`539ed6eb0d`](https://github.com/datrs/flat-tree/commit/539ed6eb0de7a1939c1c2ab65444cf23ecc47049)] Run CI on master branch instead of main (Timo Tiuraniemi)
- [[`694c6ef688`](https://github.com/datrs/flat-tree/commit/694c6ef688b61eefb338dce79d07fdbfee108141)] Switch from Travis to GHA (Timo Tiuraniemi)
- [[`597c8caa29`](https://github.com/datrs/flat-tree/commit/597c8caa29a047d05b3eb34dcc942c2579e8ade0)] Fix clippy warnings (Timo Tiuraniemi)
- [[`e707e790fb`](https://github.com/datrs/flat-tree/commit/e707e790fb0219d9516055c44711488962f12c53)] Move documentation to lib.rs for easier maintainability (Timo Tiuraniemi)
- [[`179a44d8bb`](https://github.com/datrs/flat-tree/commit/179a44d8bb1af953db5e7b60e8c32fb1d49557b8)] Fix Cargo.toml, bump edition to 2021 and, because of that, version to next major 6 (Timo Tiuraniemi)
- [[`27277762f6`](https://github.com/datrs/flat-tree/commit/27277762f62bb46ec18599775f2fab8713f6e986)] Use a new version number to avoid resolution mixups (Timo Tiuraniemi)
- [[`3c669d8c70`](https://github.com/datrs/flat-tree/commit/3c669d8c70ac42c5dfc5fa0ea46b9ec2a203a5e0)] Fix duplicated slashes in contains docs (Timo Tiuraniemi)
- [[`ec7a72ef14`](https://github.com/datrs/flat-tree/commit/ec7a72ef14f3cdf7460e87976f009960f8af5370)] Add Iterator::full_tree() (Timo Tiuraniemi)
- [[`8bc607394a`](https://github.com/datrs/flat-tree/commit/8bc607394af5303c229fe43326913ffc6ce41920)] Use a bigger tree in README to ease understanding of tests. Add a few more tests. (Timo Tiuraniemi)
- [[`3acd091d7e`](https://github.com/datrs/flat-tree/commit/3acd091d7e482299a430620e0f84a04451ffc4ad)] Add Iterator::next_tree() and Iterator::prev_tree() (Timo Tiuraniemi)
- [[`7048931423`](https://github.com/datrs/flat-tree/commit/7048931423738a7e3b3318ffe3cc625249b9ea79)] Add count_leaves(), Iterator::count_nodes() and Iterator::count_leaves() (Timo Tiuraniemi)
- [[`3b136873ce`](https://github.com/datrs/flat-tree/commit/3b136873ceedb3c6af5de6c00d5e59cfbce9bc6a)] Fix README picture which had confused indexes (Timo Tiuraniemi)
- [[`2bb6bf478d`](https://github.com/datrs/flat-tree/commit/2bb6bf47811d33051f3fe632c5736b4bacbf608a)] Add a contains() method to the iterator (Timo Tiuraniemi)
- [[`b1cb6ab914`](https://github.com/datrs/flat-tree/commit/b1cb6ab914e946cf3404769a782ec4442dc28e4d)] Update changelog (Bruno Tavares)

### Stats
```diff
  .github/CODE_OF_CONDUCT.md       |  75 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 .github/CONTRIBUTING.md          |  63 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 .github/ISSUE_TEMPLATE.md        |  41 +++++++++++++++++++++++++++++++++++++++++
 .github/PULL_REQUEST_TEMPLATE.md |  21 +++++++++++++++++++++
 .github/stale.yml                |  17 +++++++++++++++++
 .github/workflows/ci.yml         |  95 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 CERTIFICATE                      |  37 +++++++++++++++++++++++++++++++++++++
 CHANGELOG.md                     |  20 ++++++++++++++++++++
 Cargo.toml                       |  11 ++++++-----
 README.md                        |  61 ++-----------------------------------------------------------
 src/iterator.rs                  | 166 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 src/lib.rs                       | 103 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------
 tests/iterator.rs                |   8 ++++----
 13 files changed, 634 insertions(+), 84 deletions(-)
```


## 2020-03-03, Version 5.0.0
### Commits
- [[`05ffe8699d`](https://github.com/mafintosh/flat-tree-rs/commit/05ffe8699d452b01152a596e5cb0a652e7915503)] (cargo-release) version 5.0.0 (Bruno Tavares)
- [[`709b2d5635`](https://github.com/mafintosh/flat-tree-rs/commit/709b2d5635c4d2f9fdad48e23d55d257294dbdd9)] Merge pull request #44 from bltavares/usize-to-u64 (Bruno Tavares)
- [[`bc9c866bdb`](https://github.com/mafintosh/flat-tree-rs/commit/bc9c866bdb2101aa892b4bf917a8b1a30be1da45)] Fix travis script (Bruno Tavares)
- [[`68f7e67ed4`](https://github.com/mafintosh/flat-tree-rs/commit/68f7e67ed4c76fc6c11bb45ea78e9ec41b3d2449)] Change from usize to u64 (Bruno Tavares)
- [[`f9ca2ef68e`](https://github.com/mafintosh/flat-tree-rs/commit/f9ca2ef68e19a861117788bb6624e9529b1d6419)] Update criterion requirement from 0.2 to 0.3 (dependabot-preview[bot])
- [[`b23328579d`](https://github.com/mafintosh/flat-tree-rs/commit/b23328579d00947d6480c3e894e18e1598905204)] Update changelog (Yoshua Wuyts)

### Stats
```diff
 .travis.yml     |  6 +++---
 CHANGELOG.md    | 15 +++++++++++++++
 Cargo.toml      |  4 ++--
 src/iterator.rs | 36 ++++++++++++++++++------------------
 src/lib.rs      | 34 +++++++++++++++++-----------------
 5 files changed, 55 insertions(+), 40 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)
- [[`017436cbae`](https://github.com/mafintosh/flat-tree-rs/commit/017436cbaec00ef57b7e8311a5f674fd0cf418e4)] add desc (Mathias Buus)
- [[`388d2445bc`](https://github.com/mafintosh/flat-tree-rs/commit/388d2445bc567148867a79f315aeee3ca492b100)] first commit (Mathias Buus)

### Stats
```diff
 Cargo.toml   |   2 +-
 rustfmt.toml |   2 +-
 src/lib.rs   | 328 ++++++++++++++++++++++++++----------------------------------
 3 files changed, 144 insertions(+), 188 deletions(-)
```


## 2016-03-09, Version v1.0.0
### Commits
- [[`b256517aef`](https://github.com/mafintosh/flat-tree-rs/commit/b256517aefb4ce2092e2920931d4c78cca6590af)] add badge (Mathias Buus)

tests/iterator.rs
extern crate flat_tree;

#[test]
fn iterator() {
  let mut iterator = flat_tree::Iterator::default();
  assert_eq!(iterator.index(), 0);
  assert_eq!(iterator.prev(), 0);
  assert!(iterator.is_left());
  assert!(!iterator.is_right());
  assert_eq!(iterator.parent(), 1);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.parent(), 3);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.parent(), 7);
  assert_eq!(iterator.offset(), 0);
  assert_eq!(iterator.right_child(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.next(), Some(13));
  assert!(!iterator.is_left());
  assert!(iterator.is_right());
  assert_eq!(iterator.left_span(), 12);
  assert_eq!(iterator.next(), Some(14));
  assert_eq!(iterator.next(), Some(16));
  assert_eq!(iterator.offset(), 8);
  assert_eq!(iterator.parent(), 17);
  assert_eq!(iterator.offset(), 4);
  assert_eq!(iterator.parent(), 19);
  assert_eq!(iterator.offset(), 2);
  assert_eq!(iterator.parent(), 23);
  assert_eq!(iterator.offset(), 1);
  assert_eq!(iterator.right_span(), 30);

  iterator.seek(23);

  assert_eq!(iterator.index(), 23);
  assert_eq!(iterator.offset(), 1);
  assert_eq!(iterator.right_child(), 27);
  assert_eq!(iterator.sibling(), 19);
  assert_eq!(iterator.prev(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.prev(), 5);
  assert_eq!(iterator.left_child(), 4);
  assert_eq!(iterator.prev(), 2);
  assert_eq!(iterator.prev(), 0);
}

#[test]
fn non_leaf_start() {
  let mut iterator = flat_tree::Iterator::new(1);
  assert_eq!(iterator.index(), 1);
  assert_eq!(iterator.parent(), 3);
  assert_eq!(iterator.parent(), 7);
  assert_eq!(iterator.right_child(), 11);
  assert_eq!(iterator.left_child(), 9);
  assert_eq!(iterator.next(), Some(13));
  assert_eq!(iterator.left_span(), 12);
}


src/lib.rs
#![forbid(missing_docs)]
#![cfg_attr(test, deny(warnings))]
#![doc(test(attr(deny(warnings))))]
//! # Series of functions to map a binary tree to a list
//!
//! You can represent a binary tree in a simple flat list using the following
//! structure:
//!
//! ```text
//!                                     15
//!               7                                             23
//!       3                 11                      19                      27
//!   1       5        9          13          17          21          25          29
//! 0   2   4   6   8    10    12    14    16    18    20    22    24    26    28    30...
//! ```
//!
//! Each number represents an **index** in a flat list. So a tree:
//!
//! ```text
//!       A
//!   B       C
//! D   E   F   G  ...
//! ```
//!
//! would be represented as a list: `[D B E A F C G]`
//!
//! Furthermore, indexes `0`, `2`, `4`, `6` are on **depth** `0`. `1`, `5`, `9` on depth `1`. And so forth.
//!
//! ```text
//! depth = 2  ^        3
//! depth = 1  |    1       5
//! depth = 0  |  0   2   4   6  ...
//! ```
//!
//! In some cases it is also useful to calculate an **offset**. Indexes `0`, `1`, `3`, `7` have an offset `0`:
//!
//! ```text
//!                 (7)
//!        (3)
//!   (1)       5
//! (0)   2   4   6      ...
//! ```
//!
//! `2`, `5`, `11`, `23` offset `1`:
//!
//! ```text
//!                  7
//!        3                  (11)
//!   1        (5)        9          13
//! 0   (2)   4   6    8    10    12    14
//! ```
//!
//! This module exposes a series of functions to help you build and maintain
//! this data structure.

mod iterator;

pub use iterator::Iterator;

/// Returns the flat-tree of the tree node at the specified depth and offset.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::index(0, 0), 0);
/// assert_eq!(flat_tree::index(0, 1), 2);
/// assert_eq!(flat_tree::index(0, 2), 4);
/// assert_eq!(flat_tree::index(1, 2), 9);
/// assert_eq!(flat_tree::index(1, 3), 13);
/// assert_eq!(flat_tree::index(2, 1), 11);
/// assert_eq!(flat_tree::index(2, 2), 19);
/// assert_eq!(flat_tree::index(3, 0), 7);
/// assert_eq!(flat_tree::index(3, 1), 23);
/// ```
#[inline]
pub const fn index(depth: u64, offset: u64) -> u64 {
  (offset << (depth + 1)) | ((1 << depth) - 1)
}

/// Returns the depth of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::depth(0), 0);
/// assert_eq!(flat_tree::depth(1), 1);
/// assert_eq!(flat_tree::depth(2), 0);
/// assert_eq!(flat_tree::depth(3), 2);
/// assert_eq!(flat_tree::depth(4), 0);
/// ```
#[inline]
pub const fn depth(i: u64) -> u64 {
  // Count trailing `1`s of the binary representation of the number.
  (!i).trailing_zeros() as u64
}

/// Returns the offset of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::offset(0), 0);
/// assert_eq!(flat_tree::offset(1), 0);
/// assert_eq!(flat_tree::offset(2), 1);
/// assert_eq!(flat_tree::offset(3), 0);
/// assert_eq!(flat_tree::offset(4), 2);
/// ```
#[inline]
pub fn offset(i: u64) -> u64 {
  let depth = self::depth(i);
  if is_even(i) {
    i / 2
  } else {
    i >> (depth + 1)
  }
}

/// Returns the parent of a node with a depth.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::parent(0), 1);
/// assert_eq!(flat_tree::parent(1), 3);
/// assert_eq!(flat_tree::parent(2), 1);
/// assert_eq!(flat_tree::parent(3), 7);
/// assert_eq!(flat_tree::parent(4), 5);
/// ```
#[inline]
pub fn parent(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth + 1, offset(i) >> 1)
}

/// Returns the sibling of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::sibling(0), 2);
/// assert_eq!(flat_tree::sibling(2), 0);
/// assert_eq!(flat_tree::sibling(1), 5);
/// assert_eq!(flat_tree::sibling(5), 1);
/// ```
#[inline]
pub fn sibling(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth, offset(i) ^ 1)
}

/// Returns the parent's sibling, of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::uncle(0), 5);
/// assert_eq!(flat_tree::uncle(2), 5);
/// assert_eq!(flat_tree::uncle(1), 11);
/// assert_eq!(flat_tree::uncle(5), 11);
/// ```
#[inline]
pub fn uncle(i: u64) -> u64 {
  let depth = self::depth(i);
  index(depth + 1, offset(parent(i)) ^ 1)
}

/// Returns both children of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::children(0), None);
/// assert_eq!(flat_tree::children(1), Some((0, 2)));
/// assert_eq!(flat_tree::children(3), Some((1, 5)));
/// assert_eq!(flat_tree::children(9), Some((8, 10)));
/// ```
#[inline]
pub fn children(i: u64) -> Option<(u64, u64)> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some((i, i))
  } else {
    let offset = offset(i) * 2;
    Some((index(depth - 1, offset), index(depth - 1, offset + 1)))
  }
}

/// Returns only the left child of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::left_child(0), None);
/// assert_eq!(flat_tree::left_child(1), Some(0));
/// assert_eq!(flat_tree::left_child(3), Some(1));
/// ```
#[inline]
pub fn left_child(i: u64) -> Option<u64> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some(i)
  } else {
    Some(index(depth - 1, offset(i) << 1))
  }
}

/// Returns only the left child of a node.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::right_child(0), None);
/// assert_eq!(flat_tree::right_child(1), Some(2));
/// assert_eq!(flat_tree::right_child(3), Some(5));
/// ```
// TODO: handle errors
#[inline]
pub fn right_child(i: u64) -> Option<u64> {
  let depth = self::depth(i);
  if is_even(i) {
    None
  } else if depth == 0 {
    Some(i)
  } else {
    Some(index(depth - 1, (offset(i) << 1) + 1))
  }
}

/// Returns the right most node in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::right_span(0), 0);
/// assert_eq!(flat_tree::right_span(1), 2);
/// assert_eq!(flat_tree::right_span(3), 6);
/// assert_eq!(flat_tree::right_span(23), 30);
/// assert_eq!(flat_tree::right_span(27), 30);
/// ```
#[inline]
pub fn right_span(i: u64) -> u64 {
  let depth = self::depth(i);
  if depth == 0 {
    i
  } else {
    (offset(i) + 1) * (2 << depth) - 2
  }
}

/// Returns the left most node in the tree that it spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::left_span(0), 0);
/// assert_eq!(flat_tree::left_span(1), 0);
/// assert_eq!(flat_tree::left_span(3), 0);
/// assert_eq!(flat_tree::left_span(23), 16);
/// assert_eq!(flat_tree::left_span(27), 24);
/// ```
#[inline]
pub fn left_span(i: u64) -> u64 {
  let depth = self::depth(i);
  if depth == 0 {
    i
  } else {
    offset(i) * (2 << depth)
  }
}

/// Returns the left and right most nodes in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::spans(0), (0, 0));
/// assert_eq!(flat_tree::spans(1), (0, 2));
/// assert_eq!(flat_tree::spans(3), (0, 6));
/// assert_eq!(flat_tree::spans(23), (16, 30));
/// assert_eq!(flat_tree::spans(27), (24, 30));
/// ```
#[inline]
pub fn spans(i: u64) -> (u64, u64) {
  (left_span(i), right_span(i))
}

/// Returns how many nodes are in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::count(0), 1);
/// assert_eq!(flat_tree::count(1), 3);
/// assert_eq!(flat_tree::count(3), 7);
/// assert_eq!(flat_tree::count(5), 3);
/// assert_eq!(flat_tree::count(7), 15);
/// assert_eq!(flat_tree::count(15), 31);
/// assert_eq!(flat_tree::count(23), 15);
/// assert_eq!(flat_tree::count(27), 7);
/// ```
#[inline]
pub const fn count(i: u64) -> u64 {
  let depth = self::depth(i);
  (2 << depth) - 1
}

/// Returns how many leaves are in the tree that the node spans.
///
/// ## Examples
/// ```rust
/// assert_eq!(flat_tree::count_leaves(0), 1);
/// assert_eq!(flat_tree::count_leaves(1), 2);
/// assert_eq!(flat_tree::count_leaves(3), 4);
/// assert_eq!(flat_tree::count_leaves(5), 2);
/// assert_eq!(flat_tree::count_leaves(15), 16);
/// assert_eq!(flat_tree::count_leaves(23), 8);
/// assert_eq!(flat_tree::count_leaves(27), 4);
/// ```
#[inline]
pub const fn count_leaves(i: u64) -> u64 {
  (count(i) + 1) / 2
}

/// Returns a list of all the full roots (subtrees where all nodes have either 2
/// or 0 children) `<` index.
///
/// For example `fullRoots(8)` returns `[3]` since the subtree rooted at `3`
/// spans `0 -> 6`, and the tree rooted at `7` has a child located at `9` which
/// is `>= 8`.
///
/// ## Panics
/// If an uneven index is passed.
///
/// ## Examples
/// ```rust
/// use flat_tree::full_roots;
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(0, &mut nodes);
/// assert_eq!(nodes, []);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(2, &mut nodes);
/// assert_eq!(nodes, [0]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(8, &mut nodes);
/// assert_eq!(nodes, [3]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(20, &mut nodes);
/// assert_eq!(nodes, [7, 17]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(18, &mut nodes);
/// assert_eq!(nodes, [7, 16]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(16, &mut nodes);
/// assert_eq!(nodes, [7]);
///
/// let mut nodes = Vec::with_capacity(16);
/// full_roots(30, &mut nodes);
/// assert_eq!(nodes, [7, 19, 25, 28]);

/// ```
#[inline]
pub fn full_roots(i: u64, nodes: &mut Vec<u64>) {
  assert!(
    is_even(i),
    "You can only look up roots for depth 0 blocks, got index {}",
    i
  );
  let mut tmp = i >> 1;
  let mut offset = 0;
  let mut factor = 1;

  loop {
    if tmp == 0 {
      break;
    }
    while factor * 2 <= tmp {
      factor *= 2;
    }
    nodes.push(offset + factor - 1);
    offset += 2 * factor;
    tmp -= factor;
    factor = 1;
  }
}

#[inline]
pub(crate) const fn is_even(num: u64) -> bool {
  (num & 1) == 0
}

#[inline]
pub(crate) const fn is_odd(num: u64) -> bool {
  (num & 1) != 0
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_is_even() {
    assert!(is_even(0));
    assert!(!is_even(1));
    assert!(is_even(2));
    assert!(!is_even(3));
  }

  #[test]
  fn test_is_odd() {
    assert!(!is_odd(0));
    assert!(is_odd(1));
    assert!(!is_odd(2));
    assert!(is_odd(3));
  }

  #[test]
  fn test_parent_gt_int32() {
    assert_eq!(parent(10_000_000_000), 10_000_000_001);
  }

  #[test]
  fn test_child_to_parent_to_child() {
    let mut child = 0;
    for _ in 0..50 {
      child = parent(child)
    }
    assert_eq!(child, 1_125_899_906_842_623);
    for _ in 0..50 {
      child = left_child(child).expect("no valid number returned");
    }
    assert_eq!(child, 0);
  }
}
